### DISCUSSION TITLE
Files from nuget package not copied to the build output

### DISCUSSION THREAD

**Discussion by aglasencnik** [VsixCommunity/Community.VisualStudio.Toolkit#474] [Q&A] [answered]
Hi, I have a probably quick question.
Just today I published my very first nuget package: [EmmetNetSharp](https://www.nuget.org/packages/EmmetNetSharp), and I want to base my Visual Studio extension on this nuget package...
But this nuget package needs to copy a folder with 6 javascript files on build.... (I have set the copy to always there). Basically when I try to use the nuget package on a console app it shows the files and automatically copies them, so it works straight away... but not in the extension.... it doesn't copy the files to the build output. And because it is a package I can't even see the files so I could check the copy to output and to vsix...
Any ideas on how to fix this?

**Comment by madskristensen** [comment]
Can you manually add them to the project and set the build action?

**Reply by aglasencnik** [reply]
How would I do that?
Oh you mean copy the whole project directorynand copy it in the solution with the extension and then reference it?
Or could it be done with a nuget package?

**Reply by madskristensen** [reply]
You could either write some MSBuild stuff in the .csproj to handle it, or simply add the 6 .js files to the project and set their Build Action to content

**Reply by madskristensen** [reply]
Here's how I"ve done it in the past
https://github.com/madskristensen/SassCompiler/blob/master/src/SassCompiler.csproj#L103

**Reply by aglasencnik** [reply]
Thank you.
Is the second example the on how to do it with MSBuild? So I would just "reference" the files there and thats it?

**Reply by madskristensen** [reply]
Yeah, you can add the .js files to a folder in the solution root and then reference them from the .csproj file manually. That's one way of doing it

**Reply by aglasencnik** [reply]
Ok, thank you so much.

**Comment by reduckted** [comment, marked as answer]
There is a way that you can do this without needing to manually copy files from the NuGet package.

On the `PackageReference` element, you can set [`GeneratePathProperty`](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#generatepathproperty) to true. That will create a property with the path to the NuGet package.

```xml
<PackageReference Include="EmmetNetSharp" Version="1.0.0" GeneratePathProperty="true"/>
```

In this instance, the property will be called `PkgEmmetNetSharp` (you can look in the `obj/*.nuget.g.props` file to confirm the name).

You can then use that property to include the files from the NuGet package. For example, if you want to include the files in the VSIX file, you can do this:

```xml
<ItemGroup>
  <Content Include="$(PkgEmmetNetSharp)\content\Scripts\*.js" IncludeInVsix="true" VsixSubPath="Scripts" />
</ItemGroup>
```

In Solution Explorer:
![image](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/assets/10321525/7b665eeb-8560-4639-9446-a17295ca7dda)

In the VSIX (viewed using 7zip):
![image](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/assets/10321525/7dc791c0-92e4-4d5f-a46d-ae5e82258c0a)


If you don't want the files to be visible in Solution Explorer, you can hide them using the `Visible` attribute:
```xml
<ItemGroup>
  <Content Include="$(PkgEmmetNetSharp)\content\Scripts\*.js" IncludeInVsix="true" VsixSubPath="Scripts" Visible="false" />
</ItemGroup>
```

Or you can put them in a virtual folder:
```xml
<ItemGroup>
  <Content Include="$(PkgEmmetNetSharp)\content\Scripts\*.js" IncludeInVsix="true" VsixSubPath="Scripts">
    <Link>Scripts\%(Filename)%(Extension)</Link>
  </Content>
</ItemGroup>
```
![image](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/assets/10321525/a9d27230-4e4f-4090-be8e-9f3a658c32fe)

**Reply by aglasencnik** [reply]
Hi, I just don't know what to say. Thank you so so much for taking your time, and even testing it, so we know that it works...  I really appreciate that. 
Thank you so much!!!
